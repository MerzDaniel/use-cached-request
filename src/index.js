
export {useCachedRequest, RequestProvider} from './use-cached-request'
export {useRequest} from './use-request'