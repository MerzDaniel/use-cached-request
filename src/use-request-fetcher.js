import React from "react";

const pendingCache = {}

export function useRequestFetcher(url, requests, setRequests, httpClient) {
  React.useEffect(() => {
    if (
      pendingCache[url]
      || (requests.data[url] && !(requests.states[url] && requests.states[url].refetch))
    ) {
      // already pending or loaded
      return
    }

    pendingCache[url] = true
    setRequests({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {
          pending: true,
          refetch: false,
        },
      },
    })

    httpClient.get(url).then(
      ({data: responseData}) => {
        pendingCache[url] = false
        setRequests({
          data: {
            ...requests.data,
            [url]: responseData,
          },
          states: {
            ...requests.states,
            [url]: {pending: false},
          },
        })
      },
    ).catch((e) => {
      pendingCache[url] = false
      setRequests({
        data: {...requests.data},
        states: {
          ...requests.states,
          [url]: {pending: false, error: e},
        },
      })
    })
  }, [url, requests.states[url]])

  function refetch() {
    setRequests({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {...requests.states[url], refetch: true},
      },
    })
  }

  return {
    error: requests.states[url] ? requests.states[url].error : null,
    isPending: requests.states[url] ? requests.states[url].pending : true,
    data: requests.data[url],
    refetch,
  }
}