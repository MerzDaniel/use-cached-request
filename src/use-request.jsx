import React from 'react'
import axios from 'axios'

const httpClient = axios
const pendingCache = {}

const useRequest = (url) => {
  const [requests, setRequestStates] = React.useState({
    states: {},
    data: {},
  })

  React.useEffect(() => {
    // already pending or loaded
    if (
      pendingCache[url]
      || (requests.data[url] && !(requests.states[url] && requests.states[url].refetch))
    ) {
      return
    }

    pendingCache[url] = true
    setRequestStates({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {
          pending: true,
          refetch: false,
        },
      },
    })

    httpClient.get(url).then(
      ({data: responseData}) => {
        pendingCache[url] = false
        setRequestStates({
          data: {
            ...requests.data,
            [url]: responseData,
          },
          states: {
            ...requests.states,
            [url]: {pending: false},
          },
        })
      },
    ).catch((e) => {
      pendingCache[url] = false
      setRequestStates({
        data: {...requests.data},
        states: {
          ...requests.states,
          [url]: {pending: false, error: e},
        },
      })
    })
  }, [url, requests.states[url]])

  function refetch() {
    setRequestStates({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {...requests.states[url], refetch: true},
      },
    })
  }

  return {
    error: requests.states[url] ? requests.states[url].error : null,
    isPending: requests.states[url] ? requests.states[url].pending : true,
    data: requests.data[url],
    refetch,
  }
}

export {
  useRequest,
}
