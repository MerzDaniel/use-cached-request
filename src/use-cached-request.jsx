import React from 'react'
import axios from 'axios'
import * as PropTypes from 'prop-types'
import {useRequestFetcher} from "./use-request-fetcher";

const httpClient = axios
const defaultState = {
  states: {},
  data: {},
}
const requestContext = React.createContext()
const RequestProvider = ({children}) => {
  const [state, setState] = React.useState(defaultState)

  return (
    <requestContext.Provider value={[state, setState]}>
      {children}
    </requestContext.Provider>
  )
}
RequestProvider.propTypes = {
  children: PropTypes.node.isRequired,
}

function useCachedRequest(url) {
  const [requests, setRequestStates] = React.useContext(requestContext) || [undefined, undefined]

  const {
    refetch, isPending, error, data
  } = useRequestFetcher(url, requests, setRequestStates, httpClient)

  return {
    refetch, isPending, error, data
  }
}

export {
  RequestProvider,
  useCachedRequest,
}
