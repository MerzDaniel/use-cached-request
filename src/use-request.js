import React from 'react'
import axios from 'axios'
import {useRequestFetcher} from "./use-request-fetcher";

const httpClient = axios

const useRequest = (url) => {
  const [requests, setRequestStates] = React.useState({
    states: {},
    data: {},
  })

  useRequestFetcher(url, requests, setRequestStates, httpClient)

  function refetch() {
    setRequestStates({
      data: requests.data,
      states: {
        ...requests.states,
        [url]: {...requests.states[url], refetch: true},
      },
    })
  }

  return {
    error: requests.states[url] ? requests.states[url].error : null,
    isPending: requests.states[url] ? requests.states[url].pending : true,
    data: requests.data[url],
    refetch,
  }
}

export {
  useRequest,
}
