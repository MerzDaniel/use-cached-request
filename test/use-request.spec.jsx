import React from 'react'
import sinon from 'sinon'
import axios from 'axios'
import {create, act} from 'react-test-renderer'
import './utils/ignore-comp-update-outside-act-warning'

import {useRequest} from '../src/use-request'

function TestComp() {
  const hook = useRequest('/some/cool/api/123')
  const {data, isPending, error} = hook
  const res = `${isPending || data || error}`
  return <div hook={hook}>
    {res}
  </div>
}

describe('tests', () => {
  let sandbox
  beforeEach(() => {
    sandbox = sinon.createSandbox()
  })
  afterEach(() => {
    sandbox.restore()
  })

  it('should show pending state', () => {
    sandbox.stub(axios, 'get').resolves({data: 'foo'})
    let comp
    act(() => {
      comp = create(<TestComp/>)
    })
    expect(comp.toJSON()).toMatchSnapshot()
  })
  it('should fetch data', (done) => {
    let stub = sandbox.stub(axios, 'get').resolves({data: 'bar'})
    let comp
    act(() => {
      comp = create(<TestComp/>)
    })
    setImmediate(() => {
      // wait for IO
      expect(comp.toJSON()).toMatchSnapshot()
      expect(stub.callCount).toBe(1)
      done()
    })
  })
})
