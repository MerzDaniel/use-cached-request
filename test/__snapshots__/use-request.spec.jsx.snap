// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for useCachedRequest should show error 1`] = `
<div
  data-request={
    Object {
      "data": undefined,
      "error": [Error: Error],
      "isPending": false,
      "refetch": [Function],
    }
  }
>
  Error: Error
</div>
`;

exports[`Test for useCachedRequest should show fetched data 1`] = `
<div
  data-request={
    Object {
      "data": "bar",
      "error": undefined,
      "isPending": false,
      "refetch": [Function],
    }
  }
>
  bar
</div>
`;

exports[`Test for useCachedRequest should show pending state 1`] = `
<div
  data-request={
    Object {
      "data": undefined,
      "error": undefined,
      "isPending": true,
      "refetch": [Function],
    }
  }
>
  true
</div>
`;

exports[`Test for useRequest should show error 1`] = `
<div
  data-request={
    Object {
      "data": undefined,
      "error": [Error: Error],
      "isPending": false,
      "refetch": [Function],
    }
  }
>
  Error: Error
</div>
`;

exports[`Test for useRequest should show fetched data 1`] = `
<div
  data-request={
    Object {
      "data": "bar",
      "error": undefined,
      "isPending": false,
      "refetch": [Function],
    }
  }
>
  bar
</div>
`;

exports[`Test for useRequest should show pending state 1`] = `
<div
  data-request={
    Object {
      "data": undefined,
      "error": undefined,
      "isPending": true,
      "refetch": [Function],
    }
  }
>
  true
</div>
`;
